src/realm/util/file.cpp

```diff
@@ -1098,7 +1100,16 @@ static void _unlock(int m_fd)
     do {
         r = flock(m_fd, LOCK_UN);
     } while (r != 0 && errno == EINTR);
-    REALM_ASSERT_RELEASE_EX(r == 0 && "File::unlock()", r, errno);
+    if (r) {
```

Comment:
So almost all good now, i'm not entirely convinced about this bit still. 

We've almost misdiagnosed this issues 9th time in a row, do i understand it right? Essentially, we tried to unlock what we didn't lock (most likely from bugreports) since it wasn't needed for Realm::convert (Realm.writeCopyTo in java/kotlin sdks). 

And since the assumption was that unlock is noop if file is not locked, we called it unconditionally every time and it was fine until new storage restrictions in Android hit and this syscall started to fail. So we were crashing apps for no reason for few years already. 

Under #5107 and realm-java#6893 it was already mentioned that users activated needed permissions, and it was still crashing. To me this case already looks irrelevant, since we're not going to hit exact same code path now. The message may still be relevant for some realm opening scenarios though. @cmelchior could you confirm this? 

Also, it seems that [sqlite3](https://github.com/sqlite/sqlite/blob/master/src/os_unix.c#L2626) essentially ignored the result of unlock call failure (although logs it still internally) on close at least. The same seems to be [true for windows](https://github.com/sqlite/sqlite/blob/master/src/os_win.c#L2546) where in most cases the return result is ignored, and it never asserts or propages further.

Exception in this case not much better that assertion. @finnschiermer shouldn't we just dump the error into logger, but essentially ignore it? Why interrupt any flow here?
@nicola-cab please, it's been a long issue if we account for very first bug reports, could you justify why do you think it is still needed here?

Reply:
We need to throw an exception for 2 reasons.
a) it is just a much better user experience, image that you are developing your app and your app crashes without any apparent reason, what would you do? ... most probably open an issue in order to try to understand what's going on.

b) this is a limitation of some SDKs, but impacts also us, they don't easily print error messages that are coming from our ASSERTIONS. Yes, there are some ways for the developer to check the system logs, but I would not rely on that, and surely we can't grab those logs from the device ourselves. 

Ignoring the error can be a viable option, but I don't think it is correct. IMO we need to merge this PR in our code.